samza-core/src/main/java/org/apache/samza/execution/JobGraphJsonGenerator.java

```diff
@@ -191,13 +216,18 @@ private void updateOperatorGraphJson(MessageStreamImpl messageStream, Traversabl
    * @param opGraph {@link org.apache.samza.execution.JobGraphJsonGenerator.OperatorGraphJson}
    * @return {@link org.apache.samza.execution.JobGraphJsonGenerator.OperatorJson}
    */
-  private OperatorJson getOrCreateOperatorJson(OperatorSpec opSpec, OperatorGraphJson opGraph) {
+  private OperatorJson buildOperatorJson(OperatorSpec opSpec, OperatorGraphJson opGraph,
+      MessageStreamImpl messageStream) {
     Map<Integer, OperatorJson> operators = opGraph.operators;
     OperatorJson opJson = operators.get(opSpec.getOpId());
     if (opJson == null) {
       opJson = new OperatorJson();
       opJson.opCode = opSpec.getOpCode().name();
       opJson.opId = opSpec.getOpId();
+      StackTraceElement stackTraceElement = messageStream.getCallerStackTrace(opSpec.getOpId());
```

Comment:
Ok, I added the sourceLocation to the operator constructor and have the getSourceLocation() in the API now. Also add the toJson() to the operatorSpec. Now we can have customized JSON for different operators. Thanks a lot for this suggestion!